src/ws/ws_manager.rs

```diff
@@ -94,27 +93,53 @@ impl WsManager {
     pub(crate) async fn new(url: String) -> Result<WsManager> {
         let stop_flag = Arc::new(AtomicBool::new(false));
 
-        let (ws_stream, _) = connect_async(url.clone())
-            .await
-            .map_err(|e| Error::Websocket(e.to_string()))?;
-
-        let (writer, mut reader) = ws_stream.split();
+        let (writer, mut reader) = Self::connect(url.as_str()).await?.split();
         let writer = Arc::new(Mutex::new(writer));
 
         let subscriptions_map: HashMap<String, Vec<SubscriptionData>> = HashMap::new();
         let subscriptions = Arc::new(Mutex::new(subscriptions_map));
         let subscriptions_copy = Arc::clone(&subscriptions);
 
         {
+            let writer = writer.clone();
             let stop_flag = Arc::clone(&stop_flag);
             let reader_fut = async move {
-                // TODO: reconnect
                 while !stop_flag.load(Ordering::Relaxed) {
                     let data = reader.next().await;
-                    if let Err(err) =
-                        WsManager::parse_and_send_data(data, &subscriptions_copy).await
-                    {
-                        error!("Error processing data received by WS manager reader: {err}");
+                    match data {
```

Comment:
nit: typically we don't use `match` for options. more idiomatic to use `if let Some(data) = data { ... } else { ... }`

Reply:
Agree for Options. Note this is on a Result, not an Option, where an action needs to be taken and destructuring for both paths. However I can change to if with an additional assignment ahead of the if branch. If there's a more idiomatic way to do with if let happy to do it. Thanks!

Reply:
I'm referring to `data` which is an Option

Reply:
Ah got it. Changed to if let.